lightning-invoice/src/payment.rs

```diff
@@ -173,10 +173,32 @@ where
 	logger: L,
 	event_handler: E,
 	/// Caches the overall attempts at making a payment, which is updated prior to retrying.
-	payment_cache: Mutex<HashMap<PaymentHash, usize>>,
+	payment_cache: Mutex<HashMap<PaymentHash, PaymentAttempts>>,
 	retry_attempts: RetryAttempts,
 }
 
+/// Storing minimal payment attempts information required for determining if a outbound payment can
+/// be retried.
+#[derive(Debug,Clone,Copy)]
+struct PaymentAttempts {
+    /// This should be >= 1 as we only insert PaymentAttempts for a PaymentHash if there is at least
+    /// one attempt.
+    counts: usize,
+    latest_attempted_at: SystemTime,
+    first_attempted_at: SystemTime
```

Comment:
Do we need to store both count and time if only one is used?

Reply:
It's not necessary but the advantage is we can have more information in error log if retry fail. related comment https://github.com/lightningdevkit/rust-lightning/pull/1418#discussion_r849143020

And if really want to optimize this, I guess it makes sense to merge payment_cache and retry_attempts fields in `InvoicePayer` into one enum field to avoid impossible state(where retry strategy is count base but we store time vice versa).

Reply:
Hmm... not sure how we'd do that because you need to know the maximum attempts when inserting into the map. Let me know if I misunderstood or if you have a more concrete idea.

We could also use a type parameter implementing a common trait (in a similar manner to the `Time` trait) such that there is never an invalid state.

Reply:
I tried but unable to do the optimization of storing only either count and time(based on Retry). and storing both has a pro for displaying user more info in retry error. So I will put this optimization on pause but please let me know if this is high-priority I can give it another try.